tests/unit/test_download.py

```diff
@@ -10,14 +11,23 @@
 
 import pip
 from pip._internal.download import (
-    PipSession, SafeFileCache, path_to_url, unpack_file_url, unpack_http_url,
-    url_to_path,
+    MultiDomainBasicAuth, PipSession, SafeFileCache, path_to_url,
+    unpack_file_url, unpack_http_url, url_to_path,
 )
 from pip._internal.exceptions import HashMismatch
 from pip._internal.models.link import Link
 from pip._internal.utils.hashes import Hashes
 from tests.lib import create_file
 
+try:
+    import keyring.backend
+    import keyring.credentials
+    from keyring.backend import KeyringBackend
+except ImportError:
+    keyring = None
+    # KeyringBackend is used as a base class
+    KeyringBackend = object
```

Comment:
These tests are being skipped without comment if the keyring package isn't present. There needs to be something to ensure that CI tests pip both with and without the keyring package installed.

If I'm reading the PR right, at the moment (1) there's nothing to get keyring installed on CI, and (2) without keyring installed, the tests get skipped. So as this stands, CI will test *neither* of the above 2 cases.

Reply:
Good point. I'll add that

Reply:
And to be clear, the tests are visibly skipped by `pytest`, so there is a message for each test, rather than a single message on import.

Reply:
Apologies, "without comment" - what I was intending to say was "without clearly explaining why the test was skipped, or what to do to avoid it being skipped". I'm not quite sure what pytest says for a skip, but at best it'll say "keyring is None". Better if it said "skipped because you don't have the keyring package installed" or something.